### PR TITLE
Add path types for cache.

### PR DESCRIPTION
**Pull request by nathanhammond**
As prework for merging cache-to-tar, this ensures that we know what types of paths we're dealing with at (almost) every step in the cache process.

It also makes the first argument to `Put` (previously unused) and `Fetch` the anchor of the enumerated list of files being saved or restored.

### CODE REVIEW

File: cli/internal/cache/cache_fs.go
Comment (nathanhammond) [inline cli/internal/cache/cache_fs.go:41]: This should have always been a `DirExists` check, not a `PathExists` check.

File: cli/internal/cache/cache_fs.go
Comment (nathanhammond) [inline cli/internal/cache/cache_fs.go:64]: This should have always been a `DirExists` check, not a `PathExists` check.

File: cli/internal/cache/cache.go
Comment (nathanhammond) [inline cli/internal/cache/cache.go:280]: With `ok` being false, we should be returning `nil` instead of the input list of `files`.

File: cli/internal/cache/cache_fs_test.go
Comment (nathanhammond) [inline cli/internal/cache/cache_fs_test.go:46]: Directories are now specified by the first argument to `Put` and `Fetch`.

File: cli/internal/cache/cache_http.go
Comment (nathanhammond) [inline cli/internal/cache/cache_http.go:281]: There are cases where this is a bug. This will be addressed in the cache-to-tar portion.

File: cli/internal/cache/cache_http.go
Comment (nathanhammond) [inline cli/internal/cache/cache_http.go:282]: This is conversion can do the wrong thing with `\`.

File: cli/internal/cache/cache_http.go
Comment (nathanhammond) [inline cli/internal/cache/cache_http.go:283]: Extra-verbose for now, will get cleaned up in cache-to-tar.

File: cli/internal/cache/cache_http_test.go
Comment (nathanhammond) [inline cli/internal/cache/cache_http_test.go:195]: Files are no longer strings! 🎉

File: cli/internal/runcache/runcache.go
Comment (nathanhammond) [inline cli/internal/runcache/runcache.go:287]: This is a known value at this point, but I think this whole loop should be cleaned up. We're doing too much work here.

File: cli/internal/runcache/runcache.go
Comment (nathanhammond) [inline cli/internal/runcache/runcache.go:290]: This looks like a material change, but previously the value passed in here was unused.

It now uses this as the anchor for the `relativePaths` slice that is getting passed in here and prevents us from depending on `cwd`.

**Review by nathanhammond — COMMENTED** [review]
A guide to the non-mechanical changes.

File: cli/internal/cache/cache_http_test.go
Comment (nathanhammond) [inline cli/internal/cache/cache_http_test.go:192]: Previously `restoreTar` was _always_ returning Unix-separated paths, even on Windows. However, downstream uses were treating these as system-separated paths. Likely didn't manifest in user-facing issues because Windows supports `/` as a separator, but no guarantees.

File: cli/internal/cache/async_cache.go
Comment (nathanhammond) [inline cli/internal/cache/async_cache.go:79]: I'm doing this to silence the lint error. It doesn't make sense unless we have an error group here and that's something that cache-to-tar will obviate the need for.

File: cli/internal/cache/cache_fs_test.go
Comment (gsoltis) [inline cli/internal/cache/cache_fs_test.go:151]: Delete commented-out code

File: cli/internal/cache/cache_fs_test.go
Comment (gsoltis) [inline cli/internal/cache/cache_fs_test.go:34]: Not blocking, but I'm really not a fan of casting for getting an instance of a path type:
 - it's hard to grep / find-usages for
 - they can be auto-inserted by the IDE
 - they don't communicate anything about why this is ok. 
 I would expect this to be something like `AbsoluteSystemPathFromUpstream`, which communicates that the author has verified that wherever the string is coming from produces absolute paths.

**Review by gsoltis — APPROVED** [review]

File: cli/internal/cache/cache_fs_test.go
Comment (nathanhammond) [reply cli/internal/cache/cache_fs_test.go:34]: The "rule" I've been following is that I've only been doing type casts in test code, and only on initialization from "raw" values (string primitives, or calls into stdlib). You'll see I didn't do casts in any application code or even in any of the more-hairy test code.

I don't feel strongly either way, so I'll defer to your opinion on this and get the casts swapped out before landing Monday.

Comment (nathanhammond) [reply cli/internal/cache/cache_fs_test.go:34]: If we're going to make this a rule we should write a custom linter that disallows all casts into `turbopath` that aren't from a string:
Screen Shot 2022-09-26 at 1 46 57 PM

(Splitting this change into a separate PR.)

Comment (nathanhammond) [reply cli/internal/cache/cache_fs_test.go:34]: Followup PR: https://github.com/vercel/turborepo/pull/2077

**Pull request merged**
